### ISSUE TITLE
False positive dead code warning

### ISSUE THREAD

**Issue by alexkazik** [rust-lang/rust#102190]
When code is generated via a proc macro there is a case which creates a false positive for the dead code warning.

Proc macro (src/lib.rs):
```rust
use proc_macro2::Span;
use quote::quote;
use syn::spanned::Spanned;
use syn::{parse_macro_input, DeriveInput, Ident};

#[proc_macro_derive(Test)]
pub fn proc_dead_code_test(tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input = parse_macro_input!(tokens as DeriveInput);
    let repr = Ident::new("i8", input.span());
    let repr_call_size = Ident::new("i8", Span::call_site());
    let vis = input.vis;
    quote! {
        impl MyEnum {
            #vis fn false_positive(self) -> #repr {
                self as i8
            }
            pub(crate) fn ok_1(self) -> #repr {
                self as i8
            }
            #vis fn ok_2(self) -> i8 {
                self as i8
            }
            #vis fn ok_3(self) -> #repr_call_size {
                self as i8
            }
        }
    }
    .into()
}
```

Test (tests/all.rs):
```rust
use proc_dead_code_test::Test;

#[allow(dead_code)]
#[derive(Test)]
#[repr(i8)]
pub(crate) enum MyEnum {
    MyValue,
}
```

The current output is:

```
warning: associated function `false_positive` is never used
 --> tests/all.rs:4:10
  |
4 | #[derive(Test)]
  |          ^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `proc_dead_code_test` (test "all") generated 1 warning
```

The generated code for all four functions are identical (except the name, obviusly), tested with `cargo expand`. 

One is reported as dead code, though no report should be given since `#[allow(dead_code)]` is used.

On 1.63.0 this is not a warning, but with 1.65.0-beta.1 it is.

I'm aware that this is a very special case and I don't mind if it doesn't get fixed but I thought I'll report it anyway.

For a quick copy-and-paste run here the Cargo.toml:
```toml
[package]
name = "proc_dead_code_test"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0.99", features = ["derive", "parsing", "extra-traits"] }
quote = "1.0.21"
proc-macro2 = "1.0.43"
```

(I don't know how to setup play.rust-lang.org to create a proc-macro and a real test, thats why it's missing.)

**Comment by mu001999** [comment]
I tested this locally and found this has been fixed, `cargo-bisect-rustc` shows:
```
********************************************************************************
Regression in nightly-2022-10-03
********************************************************************************

fetching https://static.rust-lang.org/dist/2022-10-02/channel-rust-nightly-git-commit-hash.txt
nightly manifest 2022-10-02: 40 B / 40 B [==========================================================] 100.00 % 215.77 KB/s converted 2022-10-02 to 57f097ea25f2c05f424fc9b9dc50dbd6d399845c
fetching https://static.rust-lang.org/dist/2022-10-03/channel-rust-nightly-git-commit-hash.txt
nightly manifest 2022-10-03: 40 B / 40 B [==========================================================] 100.00 % 339.06 KB/s converted 2022-10-03 to 092255976862c1f0121bb4446acd38336a860c4f
looking for regression commit between 2022-10-02 and 2022-10-03
fetching (via remote github) commits from max(57f097ea25f2c05f424fc9b9dc50dbd6d399845c, 2022-09-30) to 092255976862c1f0121bb4446acd38336a860c4f
ending github query because we found starting sha: 57f097ea25f2c05f424fc9b9dc50dbd6d399845c
get_commits_between returning commits, len: 17
  commit[0] 2022-10-01: Auto merge of #102236 - cjgillot:compute_lint_levels_by_def, r=oli-obk
  commit[1] 2022-10-01: Auto merge of #102193 - ferrocene:pa-remote-test-server-improvements, r=Mark-Simulacrum
  commit[2] 2022-10-02: Auto merge of #102558 - matthiaskrgr:rollup-0odec1c, r=matthiaskrgr
  commit[3] 2022-10-02: Auto merge of #102424 - sunfishcode:sunfishcode/hidden-main, r=nagisa
  commit[4] 2022-10-02: Auto merge of #102535 - scottmcm:optimize-split-at-partition-point, r=thomcc
  commit[5] 2022-10-02: Auto merge of #102548 - nikic:inline-cell-replace, r=scottmcm
  commit[6] 2022-10-02: Auto merge of #98354 - camsteffen:is-some-and-by-value, r=m-ou-se
  commit[7] 2022-10-02: Auto merge of #102586 - Dylan-DPC:rollup-g107h6z, r=Dylan-DPC
  commit[8] 2022-10-02: Auto merge of #102530 - cuviper:upgrade-i586-gnu, r=pietroalbini
  commit[9] 2022-10-02: Auto merge of #100557 - dawnofmidnight:tarball-commit-info, r=Mark-Simulacrum
  commit[10] 2022-10-02: Auto merge of #102305 - flba-eb:remove_exclude_list, r=Mark-Simulacrum
  commit[11] 2022-10-03: Auto merge of #102503 - cuviper:x86-stack-probes, r=nagisa
  commit[12] 2022-10-03: Auto merge of #102508 - nnethercote:even-more-lexer-improvements, r=matklad
  commit[13] 2022-10-03: Auto merge of #102614 - matthiaskrgr:rollup-fjyu7oo, r=matthiaskrgr
  commit[14] 2022-10-03: Auto merge of #102551 - bjorn3:cg_ssa_cleanup, r=davidtwco
  commit[15] 2022-10-03: Auto merge of #102610 - dawnofmidnight:git-commit-hash, r=Mark-Simulacrum
  commit[16] 2022-10-03: Auto merge of #102627 - matthiaskrgr:rollup-2xtrqkw, r=matthiaskrgr
```

I think it might be https://github.com/rust-lang/rust/pull/102538.
